File: day5/sol.py
from typing import Dict, List
from functools import cmp_to_key


def part1(rules: Dict[int, List[int]], updates: List[List[int]]) -> int:
    sum_ = sum([update[len(update)//2] for update in updates if
                not any([rules[update[j]].count(update[i]) > 0
                         for i in range(len(update))
                         for j in range(i+1, len(update))])])

    return sum_


def part2(rules: Dict[int, List[int]], original_updates: List[List[int]]) -> int:
    sorted_updates = [sorted(update, key=cmp_to_key(
        lambda x, y: -1 if x in rules and y in rules[x] else 1)) for update in original_updates]

    return part1(rules, sorted_updates) - part1(rules, original_updates)


def solve(input):
    with open(input) as file:
        lines = file.readlines()
        separator_index = lines.index("\n")
        pairs = [list(map(int, line.strip().split("|")))
                 for line in lines[:separator_index]]
        rules = {x: [y for _, y in pairs if _ == x] for x, _ in pairs}
        updates = [list(map(int, line.strip().split(",")))
                   for line in lines[separator_index+1:]]

    return (part1(rules, updates), part2(rules, updates))


if __name__ == "__main__":
    input = "input.txt"
    result = solve(input)
    print("Part 1:", result[0])
    print("Part 2:", result[1])

File: day5/sol.cpp
// AI generated translation from sol.py

#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <sstream>

using namespace std;

int part1(const unordered_map<int, vector<int>>& rules, const vector<vector<int>>& updates) {
    int sum = 0;
    for (const auto& update : updates) {
        bool valid = true;
        for (size_t i = 0; i < update.size(); ++i) {
            for (size_t j = i + 1; j < update.size(); ++j) {
                if (rules.at(update[j]).end() != find(rules.at(update[j]).begin(), rules.at(update[j]).end(), update[i])) {
                    valid = false;
                    break;
                }
            }
            if (!valid) break;
        }
        if (valid) {
            sum += update[update.size() / 2];
        }
    }
    return sum;
}

int part2(const unordered_map<int, vector<int>>& rules, const vector<vector<int>>& original_updates) {
    vector<vector<int>> sorted_updates = original_updates;
    for (auto& update : sorted_updates) {
        sort(update.begin(), update.end(), [&rules](int x, int y) {
            if (rules.find(x) != rules.end() && find(rules.at(x).begin(), rules.at(x).end(), y) != rules.at(x).end()) {
                return true;
            }
            return false;
        });
    }
    return part1(rules, sorted_updates) - part1(rules, original_updates);
}

pair<int, int> solve(const string& input) {
    ifstream file(input);
    string line;
    vector<pair<int, int>> pairs;
    vector<vector<int>> updates;
    unordered_map<int, vector<int>> rules;

    while (getline(file, line) && !line.empty()) {
        stringstream ss(line);
        string token;
        vector<int> pair;
        while (getline(ss, token, '|')) {
            pair.push_back(stoi(token));
        }
        pairs.emplace_back(pair[0], pair[1]);
    }

    for (const auto& p : pairs) {
        rules[p.first].push_back(p.second);
    }

    while (getline(file, line)) {
        stringstream ss(line);
        string token;
        vector<int> update;
        while (getline(ss, token, ',')) {
            update.push_back(stoi(token));
        }
        updates.push_back(update);
    }

    return {part1(rules, updates), part2(rules, updates)};
}

int main() {
    string input = "input.txt";
    auto result = solve(input);
    cout << "Part 1: " << result.first << endl;
    cout << "Part 2: " << result.second << endl;
    return 0;
}